services/github/src/handlers.js

```diff
@@ -630,19 +630,42 @@ async function jobHandler(message) {
     if (!defaultBranchYml) { return; }
 
     if (this.getRepoPolicy(defaultBranchYml).startsWith('collaborators')) {
-      let login = message.payload.body.pull_request.head.user.login;
-
-      let isCollaborator = await instGithub.repos.checkCollaborator({
-        owner: organization,
-        repo: repository,
-        username: login,
-      }).catch(e => {
-        if (e.status !== 404) {
-          throw e;
-        }
-      });
-
-      if (!isCollaborator) {
+      // There are four usernames associated with a PR action:
+      //  - pull_request.user.login -- the user who opened the PR
+      //  - pull_request.head.user.login -- the username or org name for the repo from which changes are pulled
+      //  - pull_request.base.user.login -- the username or org name for the repo into which changes will merge
+      //  - sender.login -- the user who clicked the button to trigger this action
+      //
+      // The "collaborators" and "collaborators_quiet" policies require:
+      //  - pull_request.user.login is a collaborator; AND
+      //  - pull_request.head.user.login is
+      //    - a collaborator OR
+      //    - the same as pull_request.base.user.login
+      //
+      // Meaning that the PR must have been opened by a collaborator and be merging code from a collaborator
+      // or from the repo against which the PR is filed.
+
+      const isCollaborator = async login => {
+        return Boolean(await instGithub.repos.checkCollaborator({
+          owner: organization,
+          repo: repository,
+          username: login,
+        }).catch(e => {
+          if (e.status !== 404) {
+            throw e;
+          }
+          return false; // 404 -> false
+        }));
+      };
+
+      const evt = message.payload.body;
+      const opener = evt.pull_request.user.login;
+      const openerIsCollaborator = await isCollaborator(opener);
+      const head = evt.pull_request.head.user.login;
+      const headIsCollaborator = head === opener ? openerIsCollaborator : await isCollaborator(head);
```

Comment:
why `opener ?` check? We already used `opener` by this point, so `openerIsCollaborator && await ...` should be enough?

Reply:
It's a little bit of an optimization to call the GitHub API less often.  The reasoning is that the common case is that the head and the opener are the same user, so in that case (`head === opener`) we just use the result of the `isCollaborator` call for that user (`openerIsCollaborator`).  Otherwise, we call the GH API again for the head user.